api/v1/sp.go

```diff
@@ -1,73 +1,117 @@
+//	@Param			body StorageProviderIDs true "List of SP IDs to approve"
 //	@Produce		json
 //	@Success		200	{object}	ResponseEnvelope{response=GetStorageProvidersResponse}
 //	@Router			/sp/approve [post]
-func handleApproveStorageProviders(c echo.Context) error {
-	return c.JSON(http.StatusNotImplemented, map[string]string{})
+func (a *apiV1) handleApproveStorageProviders(c echo.Context) error {
+	storageProviderIds, err := a.updateTenantSpState(c, db.TenantSpStatePending, db.TenantSpStateActive)
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, CreateErrorResponseEnvelope(c, http.StatusInternalServerError, err.Error()))
+	}
+
+	return c.JSON(http.StatusOK, CreateSuccessResponseEnvelope(c, storageProviderIds))
 }
 
 // handleSuspendStorageProviders godoc
 //
 //	@Summary		Suspend storage providers
-//	@Description Note: This is only required if auto_suspend is false, as manual suspension is required
-//	@Security apiKey
-//	@Param body body StorageProviderIDs true "List of SP IDs to suspend"
+//	@Description	Note: This is only required if auto_suspend is false, as manual suspension is required
+//	@Security		apiKey
+//	@Param			body StorageProviderIDs true "List of SP IDs to suspend"
 //	@Produce		json
 //	@Success		200	{object}	ResponseEnvelope{response=GetStorageProvidersResponse}
 //	@Router			/sp/suspend [post]
-func handleSuspendStorageProviders(c echo.Context) error {
-	return c.JSON(http.StatusNotImplemented, map[string]string{})
+func (a *apiV1) handleSuspendStorageProviders(c echo.Context) error {
+	storageProviderIds, err := a.updateTenantSpState(c, db.TenantSpStateActive, db.TenantSpStateSuspended)
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, CreateErrorResponseEnvelope(c, http.StatusInternalServerError, err.Error()))
+	}
+
+	return c.JSON(http.StatusOK, CreateSuccessResponseEnvelope(c, storageProviderIds))
 }
 
 // handleUnsuspendStorageProvider godoc
 //
 //	@Summary		Unsuspend a storage provider
-//	@Security apiKey
-//	@Param body body StorageProviderIDs true "List of SP IDs to unsuspend"
+//	@Security		apiKey
+//	@Param			body StorageProviderIDs true "List of SP IDs to unsuspend"
 //	@Produce		json
 //	@Success		200	{object}	ResponseEnvelope{response=GetStorageProvidersResponse}
 //	@Router			/sp/unsuspend [post]
-func handleUnsuspendStorageProvider(c echo.Context) error {
-	return c.JSON(http.StatusNotImplemented, map[string]string{})
+func (a *apiV1) handleUnsuspendStorageProvider(c echo.Context) error {
+	storageProviderIds, err := a.updateTenantSpState(c, db.TenantSpStateSuspended, db.TenantSpStateActive)
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, CreateErrorResponseEnvelope(c, http.StatusInternalServerError, err.Error()))
+	}
+
+	return c.JSON(http.StatusOK, CreateSuccessResponseEnvelope(c, storageProviderIds))
+}
+
+func (a *apiV1) updateTenantSpState(c echo.Context, stateToCheck db.TenantSpState, stateToSet db.TenantSpState) ([]int, error) {
```

Comment:
I like the idea of abstracting this into a function, but I think the exact implementation could be better.

The problem is conceptually this thing should be a "db" operation, not an "api" operation - it doesn't _really_  make sense to be a receiver function on `apiV1`, but should be coupled to the database itself. I think the cleanest way to do this is actually to add a Gorm `BeforeSave`[ lifecycle hook ](https://gorm.io/docs/hooks.html#Object-Life-Cycle) to do the validation - so you can bake this logic into the `db` package i.e under `func (*TenantSPs) beforeSave()`

Once you do this, you don't need this `updateTenantSpState` function - you can make it similar to the other API handlers where you just do a DB operation, and then have the error returned if you try to do something invalid

Reply:
(I haven't actually used Gorm lifecycle hooks before, just read about them. Interested to see exactly how they work)

Reply:
when you get back from break we can decided between this and https://github.com/data-preservation-programs/spade-tenant/pull/19